### ISSUE TITLE
Ice Restart Local And Remote Video Always Shows "Not connected"

### ISSUE THREAD

**Issue by JaspalK** [webrtc/samples#863]
Chrome version : 57.0.2970.0 / dev
Chrome Device: Ubuntu 14.04, Windows 10

Steps to reproduce:
Precondition: Have the JS console open 

1. Go to https://webrtc.github.io/samples/src/content/peerconnection/restart-ice/
2. Confirm that both the local (PC1) and remote windows (PC2) do not show any video, and the message below says "Not connected" 
3. Click "Start" and wait till local video PC1 (in left window) appears
4. Click "Call" 
- Confirm that the remote video PC2 appears
5. Note both the addresses of PC1 and PC2, as seen below the respective windows
6. Click "Restart ICE"
7. Now, note the addresses of PC1 and PC2
8. Press "Hang up"

Expected results:
1. No JS console errors should be seen
2. The last 5 digits in the address of PC1 from step 5, before the ICE restart, should be different from the address of PC1 from step 7 
3. PC2 address should remain the same before/after ICE restart
4. After Hang up, only PC1 has video

Actual results:
Local and remote videos. i.e PC1 & PC2, always show "Not connected" after Step 4 and subsequent steps

Discussing with jansson@ he thinks, maybe, the newer Adapter 3.0 release could have caused this. 

[restart_ice_not_conencted_1.1.33.txt](https://github.com/webrtc/samples/files/716260/restart_ice_not_conencted_1.1.33.txt)

**Comment by fippo** [comment]
Right. I suspect this is mostly because of the maplike change where we don't support Object.keys anymore. I'll make a PR over the weekend!

**Comment by cvranjani** [comment]
This issue also exists on Chrome OS M57 57.0.2984.0 / 9199.0.0 dev

**Comment by JaspalK** [comment]
The problem is only partially fixed.  

PC2 always shows "Not connected" i.e no address for 'Remote' video

Steps to reproduce:

    Follow the steps in the the first post in this bug.

Expected results:

    Step 5: Note both the addresses of PC1 and PC2, as seen below the respective windows


Actual results:

   Step 5: Only an address for PC1 is shown.  PC2 always shows "Not connected"

[ice_restart-PC2-not-connected.txt](https://github.com/webrtc/samples/files/732034/ice_restart-PC2-not-connected.txt)

**Comment by JaspalK** [comment]
The problem is only partially fixed.

PC2 always shows "Not connected" i.e no address for 'Remote' video

Steps to reproduce:

Follow the steps in the the first post in this bug.
Expected results:

Step 5: Note both the addresses of PC1 and PC2, as seen below the respective windows
Actual results:

Step 5: Only an address for PC1 is shown. PC2 always shows "Not connected"

[ice_restart-PC2-not-connected.txt](https://github.com/webrtc/samples/files/732060/ice_restart-PC2-not-connected.txt)

**Comment by fippo** [comment]
@JaspalK: are you trying in canary with the experimental web platform flags on?
I get some flakiness here but it works just fine in stable and M53... @henbos wdyt?

**Comment by JaspalK** [comment]
Fippo:  Opps forget to add that info.  

No, I'm trying in Chrome Dev (on windows and linux), M 57 (57.0.2987.6) with the "Experimental Web Platform features" flag enabled (on)

**Comment by JaspalK** [comment]
didn't mean to close

**Comment by fippo** [comment]
Right. That means we might have an issue in the new getStats here.

**Comment by henbos** [comment]
The code in this project is using the spec-compliant API for getStats but expecting the callback-based non-spec compliant Chrome-specific stats objects. As such, the stats it is looking for does not exist. Examples:

https://github.com/webrtc/samples/blob/d6e7e2501788848bbb20d16c7a586f3780d2bbfe/src/content/peerconnection/bandwidth/js/main.js#L224
https://github.com/webrtc/samples/blob/83f24e75187bec346b9ef4ad5654930708308e12/src/content/peerconnection/constraints/js/main.js#L211
https://github.com/webrtc/samples/blob/d6e7e2501788848bbb20d16c7a586f3780d2bbfe/src/content/peerconnection/audio/js/main.js#L275
https://github.com/webrtc/samples/blob/d6e7e2501788848bbb20d16c7a586f3780d2bbfe/src/content/datachannel/filetransfer/js/main.js#L267
https://github.com/webrtc/samples/blob/d6e7e2501788848bbb20d16c7a586f3780d2bbfe/src/content/peerconnection/restart-ice/js/main.js#L239
https://github.com/webrtc/samples/blob/73e82617ad530609236ac81ce7a859d651f5d8c1/src/content/peerconnection/audio/js/test.js#L50

**Comment by henbos** [comment]
I spoke to soon, some of those links should work with spec-compliant stats. Here's a slightly closer look.

https://github.com/webrtc/samples/blob/83f24e75187bec346b9ef4ad5654930708308e12/src/content/peerconnection/constraints/js/main.js#L220
The bytesReceived belongs to a report of type inbound-rtp and is of type [RTCInboundRTPStreamStats](https://w3c.github.io/webrtc-stats/#inboundrtpstats-dict*).
The frameWidth and frameHeight belong to [RTCMediaStreamTrackStats](https://w3c.github.io/webrtc-stats/#dom-rtcmediastreamtrackstats) of the track stats pointed to by report's [mediaTrackId](https://w3c.github.io/webrtc-stats/#dom-rtcrtpstreamstats-mediatrackid), no "googFrameHeightReceived". Though this looks like an "is video" check, which can be tested with report.mediaType == 'video'.

https://github.com/webrtc/samples/blob/83f24e75187bec346b9ef4ad5654930708308e12/src/content/peerconnection/constraints/js/main.js#L246
If a connection is the activeConnection is not specified in candidate-pair but rather an [RTCTransportStats](https://w3c.github.io/webrtc-stats/#transportstats-dict*) pointing to the candidate-pair knows if it is the active connection. The transport had "activeConnection", but this is being replaced by "dtlsState" being "connected" ([rawgit](http://rawgit.com/w3c/webrtc-stats/master/webrtc-stats.html#dom-rtctransportstats-dtlsstate)).

https://github.com/webrtc/samples/blob/83f24e75187bec346b9ef4ad5654930708308e12/src/content/peerconnection/constraints/js/main.js#L256
"ipAddress" and "portNumber" are now called "ip" and "port".

https://github.com/webrtc/samples/blob/d6e7e2501788848bbb20d16c7a586f3780d2bbfe/src/content/datachannel/filetransfer/js/main.js#L267
It's 'candidate-pair' not 'googCandidatePair', similar comment to above about active connection.

https://github.com/webrtc/samples/blob/d6e7e2501788848bbb20d16c7a586f3780d2bbfe/src/content/peerconnection/restart-ice/js/main.js#L239
Same comments as before.

If these things are fixed we have to handle all cases (both old and new stats) since we don't know which stats types the browser will provide...

**Comment by fippo** [comment]
@henbos can you copy that over to a new issue about updating all the samples to work with both styles? 

I'll add the ip / port here, I think that is the root cause at least for this demo!

**Comment by fippo** [comment]
@henbos: uh, this is bad for canary with web platform features enabled:
```
ncaught (in promise) TypeError: Failed to execute 'getStats' on 'RTCPeerConnection': The callback provided as parameter 1 is not a function.
    at RTCPeerConnection.getStats (adapter-latest.js:733)
    at onIceStateChange (main.js:239)
    at RTCPeerConnection.pc2.oniceconnectionstatechange (main.js:129)
```
We're calling the new getStats with null here? And indeed:
```
pc1.getStats(null)
```
returns an error. I'd expect this to work?

**Comment by henbos** [comment]
Uh oh, there is active discussions about what to do with the selector argument and in the meantime only the parameterless version of getStats is behind the flag. There will likely be a getStats that takes a selector argument, but it is unclear if this will be the case or if it will be of type MediaStreamTrack or something else (such as RTPSender/RTPReceiver).

Perhaps adapter.js should in the case of new getStats take a selector argument but ignore it, invoking the parameterless version? This would allow getStats(null) to work and _probably_ be a valid call for a possible future selector version.

**Comment by alvestrand** [comment]
(null) and () should always behave the same - at least according to the gospel according to @jan-ivar, I think.

**Comment by henbos** [comment]
I filed #869

**Comment by henbos** [comment]
I guess both getStats are available through the new adapter, the new one (parameterless) and the old one (function as parameter 1) based on number of arguments used to call getStats. It needs to overwrite both for the (null) and () to work as expected. Discussion to be continued in https://github.com/webrtc/adapter/pull/418

**Comment by cvranjani** [comment]
Observed this issue (Ice restart local and remote video shows "Not connected") on Chrome OS with M58 58.0.3015.0 / 9310.0.0 dev but not with M57 57.0.2987.75 / 9202.37.0 beta

**Comment by cvranjani** [comment]
@KaptenJansson : Can you please confirm if this issue is fixed ? I still see this issue on Chrome OS with M58 58.0.3029.40 / 9334.24.0 beta

**Comment by fippo** [comment]
nobody has tackled #869 yet :-/

**Comment by KaptenJansson** [comment]
Yes, AppRTC and some sample pages need an overhaul to support the new stats....Will see if I can allocate some time this quarter.

**Comment by rantonysamy** [comment]
Check you can fix this, and also add more Information along with the Connection status.

**Comment by ErikHellman** [comment]
This issue is quite old now and I'm not sure if it is valid anymore.
@JaspalK Do you mind checking it again and see if we can close this?

**Comment by rantonysamy** [comment]
He doesn't work here anymore. 

The Issue had been fixed any way, We can request an Admin to close this issue.
